File: Assets/Scripts/MazeGen/ClusterMazeGen.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using Random = UnityEngine.Random;

public class ClusterMazeGen
{

    int dimensionX;
    int dimensionY;

    ClusterMaze maze;
    private List<Cluster> lowEntropyList;

    public ClusterMazeGen()
    {
        // Default constructor
    }

    public ClusterMazeGen(int dimensionX = 10, int dimensionY = 10)
    {
        this.dimensionX = dimensionX;
        this.dimensionY = dimensionY;
    }


    public void InitMaze(int width, int height)
    {
        Debug.Log("Initializing maze with dimensions: " + width + "x" + height);
        string debugoutput;

        lowEntropyList = new List<Cluster>();

        this.dimensionX = width;
        this.dimensionY = height;

        maze = new ClusterMaze(dimensionX, dimensionY);

        // create clusters - add all possible shapes
        for (int x = 0; x < dimensionX; x++)
        {
            debugoutput = "";
            for (int y = 0; y < dimensionY; y++)
            {
                Cluster cluster = new Cluster();
                maze.SetCluster(new Coordinate(x, y), cluster);
                cluster.Location = new Coordinate(x, y);
                debugoutput += cluster.Entropy.ToString() + " ";
            }
            Debug.Log(debugoutput);
        }
        Debug.Log("=================");

        for (int x = 0; x < dimensionX; x++)
        {
            for (int y = 0; y < dimensionY; y++)
            {
                Cluster neighbor;
                Cluster cluster;
                maze.TryGetCluster(new Coordinate(x, y), out cluster);

                maze.TryGetCluster(new Coordinate(cluster.Location.X, cluster.Location.Y+1), out neighbor);
                cluster.SetNeighbor(neighbor, ClusterDirection.North);
                maze.TryGetCluster(new Coordinate(cluster.Location.X + 1, cluster.Location.Y), out neighbor);
                cluster.SetNeighbor(neighbor, ClusterDirection.East);
                maze.TryGetCluster(new Coordinate(cluster.Location.X, cluster.Location.Y-1), out neighbor);
                cluster.SetNeighbor(neighbor, ClusterDirection.South);
                maze.TryGetCluster(new Coordinate(cluster.Location.X - 1, cluster.Location.Y), out neighbor);
                cluster.SetNeighbor(neighbor, ClusterDirection.West);
            }
        }
        // start by creating a boundary around the outer edge of the maze
        // and decide where the exists will be



        //  InitializeBoundaryClusters(); bugbug maybe lets just try setting a list of concrete constraints
        // place exits and collapse
        // place any empty cells and collapse (this forced dead areas in the map - useful for making non-rectangular mazes, or areas for other visual props to exist?)
        PlaceExits();

        InitializeBoundaryClusters();

        // now pick one of the nodes with lowest entropy and collapse it
        // and propagate the constraints    
        // repeat until all nodes are collapsed or a contradiction is found
        // if a contradiction is found, restart the process
        // if the process takes too long, restart the process
        // once the maze is fully collapsed, return the maze


        UpdateLowEntropyList();

        while (UpdateLowEntropyList() == true)
        {
            // randomly pick a low entropy cell, and collapse it
            int idx = Random.Range(0, lowEntropyList.Count);
            Debug.Log("collapse at " + idx);
            Cluster target = lowEntropyList[idx];
            target.SetRandomDefiniteShape();
            PropagateConstraints(target);
        }


        Debug.Log("Final maze:");
        for (int x = 0; x < dimensionX; x++)
        {
            debugoutput = "";
            for (int y = 0; y < dimensionY; y++)
            {
                Cluster debugcluster;
                maze.TryGetCluster(new Coordinate(x, y), out debugcluster);
                debugoutput += debugcluster.DefiniteShape.ToString() + " ";
            }
            Debug.Log(debugoutput);
        }

    }

    void PropagateConstraints(Cluster cluster)
    {
        /*
        while (stack.is_empty() == False):
            tile = stack.pop()
            tilePossibilities = tile.getPossibilities()
            directions = tile.getDirections()

            for direction in directions:
                neighbour = tile.getNeighbour(direction)
                if neighbour.entropy != 0:
                    reduced = neighbour.constrain(tilePossibilities, direction)
                    if reduced == True:
                        stack.push(neighbour)    # When possibilities were reduced need to propagate further
        */

        Stack<Cluster> stack = new Stack<Cluster>();
        stack.Push(cluster);

        while (stack.Count > 0)
        {

            Cluster targetCluster = stack.Pop();

            foreach (ClusterDirection dir in ClusterDirection.GetValues(typeof(ClusterDirection)))
            {
                // do each of 4 directions in order

                Cluster neighbor = targetCluster.Neighbor(dir);
                bool modified = false;

                if (neighbor == null)
                    continue;

                // 1. get list of all current possible shapes in this cluster, 
                // 2. and for each of those shapes, get the list of compatible neighbor shapes in that direction
                // 3. then, subtract this list from the list of all shapes
                // 4. this will leave just the list of shapes that are incompatible with ANY of our possible shapes, in that direction

                List<ClusterShape> compatibleShapes = new List<ClusterShape>();
                foreach (ClusterShape possibleShape in targetCluster.PossibleShapes)
                {
                    ClusterShape[] validNeigborShapes = ClusterRules.validNeighbors[new Tuple<ClusterShape, ClusterDirection>(possibleShape, dir)];
                    if (validNeigborShapes == null)
                    {
                        continue;
                    }

                    foreach (ClusterShape neighborShape in validNeigborShapes)
                    {
                        if (!compatibleShapes.Contains(neighborShape))
                        {
                            compatibleShapes.Add(neighborShape);
                        }
                    }
                }

                // now we have all compatible shapes in this direction
                // so now remove these from the total list of shapes to reduce down to just incompatible shapes
                List<ClusterShape> remainderShapes = new List<ClusterShape>();
                foreach (ClusterShape candidateShape in ClusterShape.GetValues(typeof(ClusterShape)))
                {
                    if (!remainderShapes.Contains(candidateShape))
                    {
                        remainderShapes.Add(candidateShape);
                    }
                }

                // and remove any shape from the total list if it's a compatible shape (i.e. we done want to remove those shapes from the neighbor)
                foreach (ClusterShape currentShape in compatibleShapes)
                {
                    if (remainderShapes.Contains(currentShape))
                    {
                        remainderShapes.Remove(currentShape);
                    }

                }

                // now remainder is just the list of incompatible shapes

                if (neighbor.Entropy > 1) // > 2 since we need 1 concrete shape and the "empty" entry
                {
                    modified = neighbor.ReducePossibleShapes(remainderShapes.ToArray());
                }

                if (modified)
                {
                    stack.Push(neighbor);
                }
            }
        }

    }

   

    private bool UpdateLowEntropyList()
    {
        bool isList = false;

        int lowest = int.MaxValue;
        lowEntropyList.Clear();

        for (int i = 0; i < dimensionX; i++)
        {
            for (int j = 0; j < dimensionY; j++)
            {
                Cluster cluster;
                
                bool foundCluster = maze.TryGetCluster(new Coordinate(i, j), out cluster);

                if (foundCluster)
                {
                    if ((cluster.Entropy == 1) || (cluster.Entropy > lowest))  // tile is done or entropy is greater
                        continue;


                    if (cluster.Entropy < lowest)
                    {

                        lowest = cluster.Entropy;
                        lowEntropyList.Clear();
                        lowEntropyList.Add(cluster);
                        isList = true;

                    }
                    else if (cluster.Entropy == lowest)
                    {
                        lowEntropyList.Add(cluster);
                        isList = true;
                    }
                }

            }
        }

        return isList;
    }



    // bugbug make reference counts of each incompatible shape?
    private void CompareShapes(Cluster updatedGroup, Coordinate nextCoordinate, ClusterDirection directionToNextGroup, Cluster[,] tileGroupMatrix)
    {
      
    }

    void PlaceExits()
    {
        int exitFlags = Random.Range(1, 16);
        Debug.Log("Exit flags: " + exitFlags);

        if ((exitFlags & 1) == 1)
        { 
            // place an exit
            Cluster exitCluster;
            maze.TryGetCluster(new Coordinate(0, dimensionY / 2), out exitCluster);
            if (exitCluster == null)
            {
                Debug.LogError("Cluster at " + 0 + "," + dimensionY / 2 + " is null");
            }
            else
            {
                exitCluster.SetDefiniteShape(ClusterShape.HorizontalLine);
            }
            PropagateConstraints(exitCluster);

        }

        if ((exitFlags & 2) == 2)
        {
            // place an exit
            Cluster exitCluster;
            maze.TryGetCluster(new Coordinate(dimensionX - 1, dimensionY / 2), out exitCluster);
            if (exitCluster == null)
            {
                Debug.LogError("Cluster at " + (dimensionX - 1) + "," + dimensionY / 2 + " is null");
            }
            else
            {
                exitCluster.SetDefiniteShape(ClusterShape.HorizontalLine);
            }
            PropagateConstraints(exitCluster);

        }

        if ((exitFlags & 4) == 4)
        {
            // place an exit
            Cluster exitCluster;
            maze.TryGetCluster(new Coordinate(dimensionX / 2, 0), out exitCluster);
            if (exitCluster == null)
            {
                Debug.LogError("Cluster at " + (dimensionX / 2) + "," + 0 + " is null");
            }
            else
            {
                exitCluster.SetDefiniteShape(ClusterShape.VerticalLine);
            }
            PropagateConstraints(exitCluster);

        }

        if ((exitFlags & 8) == 8)
        {
            // place an exit
            Cluster exitCluster;
            maze.TryGetCluster(new Coordinate(dimensionX / 2, dimensionY - 1), out exitCluster);
            if (exitCluster == null)
            {
                Debug.LogError("Cluster at " + dimensionX / 2 + "," + (dimensionY - 1) + " is null");
            }
            else
            {
                exitCluster.SetDefiniteShape(ClusterShape.VerticalLine);
            }
            PropagateConstraints(exitCluster);

        }
    }

    void InitializeBoundaryClusters()
    {
        string debugoutput;
        Cluster cluster;
        maze.TryGetCluster(new Coordinate(0, 0), out cluster); // bottom left
        {
            cluster.ReducePossibleShapes(ClusterRules.ConnectsWest);
            cluster.ReducePossibleShapes(ClusterRules.ConnectsSouth);
            cluster.ReducePossibleShapes(ClusterRules.Empty);
        }
        PropagateConstraints(cluster);
        maze.TryGetCluster(new Coordinate(0, dimensionY-1), out cluster); // top left
        {
            cluster.ReducePossibleShapes(ClusterRules.ConnectsNorth);
            cluster.ReducePossibleShapes(ClusterRules.ConnectsWest);
            cluster.ReducePossibleShapes(ClusterRules.Empty);
        }
        PropagateConstraints(cluster);
        maze.TryGetCluster(new Coordinate(dimensionX-1, 0), out cluster); // bottom right
        {
            cluster.ReducePossibleShapes(ClusterRules.ConnectsSouth);
            cluster.ReducePossibleShapes(ClusterRules.ConnectsEast);
            cluster.ReducePossibleShapes(ClusterRules.Empty);
        }
        PropagateConstraints(cluster);
        maze.TryGetCluster(new Coordinate(dimensionX - 1, dimensionY - 1), out cluster); // top right
        {
            cluster.ReducePossibleShapes(ClusterRules.ConnectsNorth);
            cluster.ReducePossibleShapes(ClusterRules.ConnectsEast);
            cluster.ReducePossibleShapes(ClusterRules.Empty);
        }
        PropagateConstraints(cluster);

        for (int x = 0; x < dimensionX; x++)
        {
            for (int y = 0; y < dimensionY; y++)
            {
                // if we're on an edge, place a wall
                bool isEdge = x == 0 || x == dimensionX - 1 || y == 0 || y == dimensionY - 1;
                if (isEdge)
                {
                   
                    maze.TryGetCluster(new Coordinate(x, y), out cluster);

                    if (cluster == null)
                    {
                        Debug.LogError("Cluster at " + x + "," + y + " is null");
                        continue;
                    }

                    // place a wall
                    if (cluster.Entropy ==1)
                    {
                        continue;
                    }

                    if (x == 0 && y == 0) // bottom left corner
                    {
                        cluster.ReducePossibleShapes(ClusterRules.ConnectsWest);
                        cluster.ReducePossibleShapes(ClusterRules.ConnectsSouth);
                        cluster.ReducePossibleShapes(ClusterRules.Empty);
                    }
                    else if (x == dimensionX - 1 && y == 0) // bottom right corner
                    {
                        cluster.ReducePossibleShapes(ClusterRules.ConnectsSouth);
                        cluster.ReducePossibleShapes(ClusterRules.ConnectsEast);
                        cluster.ReducePossibleShapes(ClusterRules.Empty);
                    }
                    else if (x == 0 && y == dimensionY - 1) // Top Left corner
                    {
                        cluster.ReducePossibleShapes(ClusterRules.ConnectsNorth);
                        cluster.ReducePossibleShapes(ClusterRules.ConnectsWest);
                        cluster.ReducePossibleShapes(ClusterRules.Empty);
                    }
                    else if (x == dimensionX - 1 && (y == dimensionY - 1)) // top right corner
                    {
                        cluster.ReducePossibleShapes(ClusterRules.ConnectsNorth);
                        cluster.ReducePossibleShapes(ClusterRules.ConnectsEast);
                        cluster.ReducePossibleShapes(ClusterRules.Empty);
                    }
                    else if (x == 0) // left edge
                    {
                        cluster.ReducePossibleShapes(ClusterRules.ConnectsWest);
                        cluster.ReducePossibleShapes(ClusterRules.Empty);
                    }
                    else if (x == dimensionX - 1) // right edge
                    {
                        cluster.ReducePossibleShapes(ClusterRules.ConnectsEast);
                        cluster.ReducePossibleShapes(ClusterRules.Empty);
                    }
                    else if (y == 0) // bottom edge
                    {
                        cluster.ReducePossibleShapes(ClusterRules.ConnectsSouth);
                        cluster.ReducePossibleShapes(ClusterRules.Empty);
                    }
                    else if (y == dimensionY - 1) // top edge
                    {
                        cluster.ReducePossibleShapes(ClusterRules.ConnectsNorth);
                        cluster.ReducePossibleShapes(ClusterRules.Empty);
                    }

                    PropagateConstraints(cluster);

                }

                // randomly pick 1 to 4 possible exit sides on the boundary




                // if we're at a corner, make it a corner wall
                // if we're on an edge but not a corner, make it a straight wall    
                // if we're on an edge and at a designated exit point, make it an exit

            }
        }


        


        Debug.Log("Afer setting boundaries:");

        for (int x = 0; x < dimensionX; x++)
        {
            debugoutput = "";
            for (int y = 0; y < dimensionY; y++)
            {
                Cluster debugcluster;
                maze.TryGetCluster(new Coordinate(x, y), out debugcluster);
                debugoutput += debugcluster.Entropy.ToString() + " ";
            }
            Debug.Log(debugoutput);
        }
        Debug.Log("=================");

    }

    void CollapseAtLocation(Coordinate coordinate)
    {
        Cluster cluster;
        if (maze.TryGetCluster(coordinate, out cluster))
        {
            if (cluster.Entropy > 1)
            {
                SetRandomDefiniteShape(cluster);
                // propagate constraints to neighbors
            }
            else
            {
                // propagate constraints to neighbors
            }

        }
        else
        {
            Debug.LogError("Cannot collapse at " + coordinate + " - out of bounds");
        }
    }

    void SetRandomDefiniteShape(Cluster cluster)
    {
        if (cluster == null)
        {
            Debug.LogError("Cannot set random definite shape for null cluster");
            return;
        }

        if (cluster.IsDefiniteShapeSet)
        {
            Debug.LogWarning("Cluster already has definite shape set");
            return;
        }

        int shapeidx = Random.Range(0, cluster.PossibleShapes.Count);
        cluster.SetDefiniteShape(cluster.PossibleShapes[shapeidx]);
    }

    void UpdateNeighbors(Cluster parentCluster, Coordinate coordinate)
    {

    }

    void UpdateLowestEntropy()
    {
    }
}

File: lua/clustermazegen.lua
require 'Cluster'
require 'Coordinate'
require 'ClusterRules'
require 'ClusterMaze'

ClusterMazeGen = {}
ClusterMazeGen.__index = ClusterMazeGen


function ClusterMazeGen:new(dX, dY)
    local instance = setmetatable({}, ClusterMazeGen)
    print("new GameMazeGen")
    instance.dimX = dX
    instance.dimY = dY
    instance.lowEntropyList = {}
    return instance
end

function ClusterMazeGen:foo()
    print("foo called")
end

function ClusterMazeGen:InitMaze(width, height)
    self.lowEntropyList = {}
    self.dimX = width
    self.dimY = height
    self.maze = ClusterMaze:new(self.dimX, self.dimY)

    for x = 0, self.dimX - 1 do
        for y = 0, self.dimY - 1 do
            local cluster = Cluster:new()
            self.maze:SetCluster(Coordinate:new(x, y), cluster)
            cluster.Location = Coordinate:new(x, y)
        end
    end

    for x = 0, self.dimX - 1 do
        for y = 0, self.dimY - 1 do
            local neighbor
            local cluster
            cluster = self.maze:TryGetCluster(Coordinate:new(x, y))

            neighbor = self.maze:TryGetCluster(Coordinate:new(cluster.Location.x, cluster.Location.y + 1))
            cluster:SetNeighbor(neighbor, ClusterDirection.North)
            neighbor = self.maze:TryGetCluster(Coordinate:new(cluster.Location.x + 1, cluster.Location.y))
            cluster:SetNeighbor(neighbor, ClusterDirection.East)
            neighbor = self.maze:TryGetCluster(Coordinate:new(cluster.Location.x, cluster.Location.y - 1))
            cluster:SetNeighbor(neighbor, ClusterDirection.South)
            neighbor = self.maze:TryGetCluster(Coordinate:new(cluster.Location.x - 1, cluster.Location.y))
            cluster:SetNeighbor(neighbor, ClusterDirection.West)
        end
    end

    self:PlaceExits()
    self:InitializeBoundaryClusters()
    self:UpdateLowEntropyList()

    while self:UpdateLowEntropyList() do
        local idx = math.random(1, #self.lowEntropyList)
        local target = self.lowEntropyList[idx]
        target:SetRandomDefiniteShape()
        self:PropagateConstraints(target)
    end
end

function ClusterMazeGen:PropagateConstraintsRefCountMethod(cluster)
    local stack = {}
    table.insert(stack, cluster)

    while #stack > 0 do
        local targetCluster = table.remove(stack)

        for _, dir in ipairs(ClusterDirection:GetValues()) do
            local neighbor = targetCluster:Neighbor(dir)
            local modified = false

            if neighbor == nil then
                goto continue
            end

            local refCounts = {}
            local refs = 0
            ::continue::
        end
    end
end

function ClusterMazeGen:PropagateConstraints(cluster)
    local stack = {}
    table.insert(stack, cluster)

    while #stack > 0 do
        local targetCluster = table.remove(stack)

        for _, dir in ipairs(ClusterDirection:GetValues()) do
            local neighbor = targetCluster:Neighbor(dir)
            local modified = false

            if neighbor == nil then
                goto continue
            end

            local compatibleShapes = {}
            for _, possibleShape in ipairs(targetCluster.PossibleShapes) do
                local validNeighborShapes = ClusterRules.validNeighbors[Tuple:new(possibleShape, dir)]
                if validNeighborShapes == nil then
                    goto continue
                end

                for _, neighborShape in ipairs(validNeighborShapes) do
                    if not table.contains(compatibleShapes, neighborShape) then
                        table.insert(compatibleShapes, neighborShape)
                    end
                end
            end

            local remainderShapes = {}
            for _, candidateShape in ipairs(ClusterShape:GetValues()) do
                if not table.contains(remainderShapes, candidateShape) then
                    table.insert(remainderShapes, candidateShape)
                end
            end

            for _, currentShape in ipairs(compatibleShapes) do
                if table.contains(remainderShapes, currentShape) then
                    table.remove(remainderShapes, currentShape)
                end
            end

            if neighbor.Entropy > 1 then
                modified = neighbor:ReducePossibleShapes(remainderShapes)
            end

            if modified then
                table.insert(stack, neighbor)
            end
            ::continue::
        end
    end
end

function ClusterMazeGen:UpdateLowEntropyList()
    local isList = false
    local lowest = math.huge
    self.lowEntropyList = {}

    for i = 0, self.dimX - 1 do
        for j = 0, self.dimY - 1 do
            local cluster
            local cluster = self.maze:TryGetCluster(Coordinate:new(i, j))

            if cluster then
                if cluster.Entropy == 1 or cluster.Entropy > lowest then
                    goto continue
                end

                if cluster.Entropy < lowest then
                    lowest = cluster.Entropy
                    self.lowEntropyList = {cluster}
                    isList = true
                elseif cluster.Entropy == lowest then
                    table.insert(self.lowEntropyList, cluster)
                    isList = true
                end
            end
            ::continue::
        end
    end

    return isList
end

function ClusterMazeGen:PlaceExits()
    local exitFlags = math.random(1, 16)

    if bit.band(exitFlags, 1) == 1 then
        local exitCluster
        exitCluster = self.maze:TryGetCluster(Coordinate:new(0, self.dimY / 2))
        if exitCluster then
            exitCluster:SetDefiniteShape(ClusterShape.HorizontalLine)
            self:PropagateConstraints(exitCluster)
        end
    end

    if bit.band(exitFlags, 2) == 2 then
        local exitCluster
        exitCluster = self.maze:TryGetCluster(Coordinate:new(self.dimX - 1, self.dimY / 2))
        if exitCluster then
            exitCluster:SetDefiniteShape(ClusterShape.HorizontalLine)
            self:PropagateConstraints(exitCluster)
        end
    end

    if bit.band(exitFlags, 4) == 4 then
        local exitCluster
        exitCluster =  self.maze:TryGetCluster(Coordinate:new(self.dimX / 2, 0))
        if exitCluster then
            exitCluster:SetDefiniteShape(ClusterShape.VerticalLine)
            self:PropagateConstraints(exitCluster)
        end
    end

    if bit.band(exitFlags, 8) == 8 then
        local exitCluster
        exitCluster =  self.maze:TryGetCluster(Coordinate:new(self.dimX / 2, self.dimY - 1))
        if exitCluster then
            exitCluster:SetDefiniteShape(ClusterShape.VerticalLine)
            self:PropagateConstraints(exitCluster)
        end
    end
end

function ClusterMazeGen:InitializeBoundaryClusters()
    local debugoutput
    local cluster

    cluster = self.maze:TryGetCluster(Coordinate:new(0, 0)) -- bottom left
    cluster:ReducePossibleShapes(ClusterRules.ConnectsWest)
    cluster:ReducePossibleShapes(ClusterRules.ConnectsSouth)
    cluster:ReducePossibleShapes(ClusterRules.Empty)
    self:PropagateConstraints(cluster)

    cluster = self.maze:TryGetCluster(Coordinate:new(0, self.dimY - 1)) -- top left
    cluster:ReducePossibleShapes(ClusterRules.ConnectsNorth)
    cluster:ReducePossibleShapes(ClusterRules.ConnectsWest)
    cluster:ReducePossibleShapes(ClusterRules.Empty)
    self:PropagateConstraints(cluster)

    cluster = self.maze:TryGetCluster(Coordinate:new(self.dimX - 1, 0)) -- bottom right
    cluster:ReducePossibleShapes(ClusterRules.ConnectsSouth)
    cluster:ReducePossibleShapes(ClusterRules.ConnectsEast)
    cluster:ReducePossibleShapes(ClusterRules.Empty)
    self:PropagateConstraints(cluster)

    cluster = self.maze:TryGetCluster(Coordinate:new(self.dimX - 1, self.dimY - 1)) -- top right
    cluster:ReducePossibleShapes(ClusterRules.ConnectsNorth)
    cluster:ReducePossibleShapes(ClusterRules.ConnectsEast)
    cluster:ReducePossibleShapes(ClusterRules.Empty)
    self:PropagateConstraints(cluster)

    for x = 0, self.dimX - 1 do
        for y = 0, self.dimY - 1 do
            local isEdge = (x == 0 or x == self.dimX - 1 or y == 0 or y == self.dimY - 1)
            if isEdge then
                cluster = self.maze:TryGetCluster(Coordinate:new(x, y))

                if cluster == nil then
                    print("Cluster at " .. x .. "," .. y .. " is null")
                    goto continue
                end

                if cluster.Entropy == 1 then
                    goto continue
                end

                if x == 0 and y == 0 then -- bottom left corner
                    cluster:ReducePossibleShapes(ClusterRules.ConnectsWest)
                    cluster:ReducePossibleShapes(ClusterRules.ConnectsSouth)
                    cluster:ReducePossibleShapes(ClusterRules.Empty)
                elseif x == self.dimX - 1 and y == 0 then -- bottom right corner
                    cluster:ReducePossibleShapes(ClusterRules.ConnectsSouth)
                    cluster:ReducePossibleShapes(ClusterRules.ConnectsEast)
                    cluster:ReducePossibleShapes(ClusterRules.Empty)
                elseif x == 0 and y == self.dimY - 1 then -- Top Left corner
                    cluster:ReducePossibleShapes(ClusterRules.ConnectsNorth)
                    cluster:ReducePossibleShapes(ClusterRules.ConnectsWest)
                    cluster:ReducePossibleShapes(ClusterRules.Empty)
                elseif x == self.dimX - 1 and y == self.dimY - 1 then -- top right corner
                    cluster:ReducePossibleShapes(ClusterRules.ConnectsNorth)
                    cluster:ReducePossibleShapes(ClusterRules.ConnectsEast)
                    cluster:ReducePossibleShapes(ClusterRules.Empty)
                elseif x == 0 then -- left edge
                    cluster:ReducePossibleShapes(ClusterRules.ConnectsWest)
                    cluster:ReducePossibleShapes(ClusterRules.Empty)
                elseif x == self.dimX - 1 then -- right edge
                    cluster:ReducePossibleShapes(ClusterRules.ConnectsEast)
                    cluster:ReducePossibleShapes(ClusterRules.Empty)
                elseif y == 0 then -- bottom edge
                    cluster:ReducePossibleShapes(ClusterRules.ConnectsSouth)
                    cluster:ReducePossibleShapes(ClusterRules.Empty)
                elseif y == self.dimY - 1 then -- top edge
                    cluster:ReducePossibleShapes(ClusterRules.ConnectsNorth)
                    cluster:ReducePossibleShapes(ClusterRules.Empty)
                end

                self:PropagateConstraints(cluster)
            end
            ::continue::
        end
    end

    print("After setting boundaries:")

    for x = 0, self.dimX - 1 do
        debugoutput = ""
        for y = 0, self.dimY - 1 do
            local debugcluster
            debugcluster = self.maze:TryGetCluster(Coordinate:new(x, y))
            debugoutput = debugoutput .. debugcluster.Entropy .. " "
        end
        print(debugoutput)
    end
    print("=================")
end

function ClusterMazeGen:CollapseAtLocation(coordinate)
    local cluster
    cluster = self.maze:TryGetCluster(coordinate)
    if cluster ~= nil then
        if cluster.Entropy > 1 then
            self:SetRandomDefiniteShape(cluster)
            -- propagate constraints to neighbors
        else
            -- propagate constraints to neighbors
        end
    else
        print("Cannot collapse at " .. tostring(coordinate) .. " - out of bounds")
    end
end

function ClusterMazeGen:SetRandomDefiniteShape(cluster)
    if cluster == nil then
        print("Cannot set random definite shape for null cluster")
        return
    end

    if cluster.IsDefiniteShapeSet then
        print("Cluster already has definite shape set")
        return
    end

    local shapeidx = math.random(1, #cluster.PossibleShapes)
    cluster:SetDefiniteShape(cluster.PossibleShapes[shapeidx])
end
